[lywich] iP

lywich opened the pull request:
# Connor, the android detective
> "Wake up Lieutenant. It's me Connor." - [Connor](https://www.youtube.com/watch?v=bn3JKpXWLvI)

Connor ensures you stay on your work and takes **pride** in doing so. 

* Text-based system
* Plug and Play
* Keeps you on task, a **~~slap~~** reminder at a time 

## Get started:

1. Download it [here](https://github.com/lywich/ip/releases/tag/v0.1)
2. Open it in command prompt 
3. Traverse to directory that this file is stored in
4. Run
```
java -jar Connor.jar 
```

## Allow Connor to keep you always working 👍 

## Features:
- [x] Add/delete/find tasks
- [ ] Manage deadline (soon!)
- [ ] Alert notifications (soon!) 

###### Command examples: 
```
Todo exercise
Deadline presentation mock up /by 2023-01-28 1800
Event Dinner /from 2023-01-28 1800 /to 2023-01-28 2000
```

###### Valid Commands:
* Todo _________
* Event _________ /from ______ /to __________ 
* Deadline ________ /by
* Hi
* Bye
* Mark _number_
* Unmark _number_
* List
* Delete _number_
* Deleteall
* find _word_

_valid date inputs are in YYYY-MM-DD 1800_
## Made in Java using OOP
```ruby
public static void main(String[] args) {
        new Connor().run();
}
```

Commit:
Add Gradle support

lywich committed:
Renamed Duke.java to Connor.java

Current project is not personalised.

Project will henceforth proceed with Connor as the chatbot's name.

* program now greets the user before asking for inputs
* program now echoes user inputs
* program recognizes a "bye" command by turning it into uppercase and exits the program

Current program fulfills Level-1.

Let's ,
* Implement a logo

lywich committed:
Added Switch and Enums to handle commands better

Current structure was too messy

Anticipate future commands to be added

Changed:
* added Enums to track valid commands (Hi, add, list, bye).
* catch IllegalArgumentException to handle invalid commands
* added TaskList.java to abstract the list management away from Connor.java.
* TaskList.java utilises LinkedList to store tasks.

Let's
* add more commands

lywich committed:
Bug fixes to "add" command in Connor.java

Current "add" command adds the whole input ("add do work") instead of ("do work") to list.

Bug has been fixed.

lywich committed:
Added Mark/Unmark function + further encapsulation for responses

The printing of messages could be further simplified to a function.

* Encapsulated the printing function to make code look neater using a separate Response.java.

Added Mark/Unmark for Level-3 using a new Task.java to encapsulate a task.

Lets's
* add TODO, Events, deadlines
* delete function

lywich committed:
Minor formatting fixes and visual improvements

Fixed the formatting and naming of certain classes to better reflect their use.

lywich committed:
Removed Enum and tweaks to compensate for removal

Enum did not felt right from a future proofing standpoint.

Remove Enum, hence:
* Tweaked Connor.java to be if-else instead
* Added TODO, DEADLINE, EVENT to replace "ADD" command

TODO, DEADLINE and EVENT extends TASK

lywich committed:
Added Automated TextUITesting capability

Currently Automated TextUITesting was not enabled

Added sample commands to test if output is producing as expected.

Let's continue to update the test everytime a new command is added.

lywich committed:
Line separator format change

Changed line separator format from CRLF to LF.

lywich committed:
Added Exceptions handling for invalid commands and tasks

Current situation: Possiblility of errors with invalid inputs

A valid input is:
* TODO: always valid
* DEADLINE: "/by" is present
* EVENT: "/from" and "/to" is present

Logic and naming functions in Connor has been changed to better reflect their function and purpose.

Let's
* handle all possible errors from now on
* Add more exceptions

lywich committed:
Added support for deleting tasks from list

Current situation: No command to delete task from list

Delete has been implemented by using remove() in the linkedlist library

Noticed a possible bug with exception handling, allowing " " to be considered as valid taskName.

Let's fix the above bug

lywich committed:
Fixed bugs with regards to exception handling

Current situation: invalid commands were not being detected

Commands that are now detected as invalid:
* "todo "
* "event "
* "deadline "

lywich committed:
Abstracted logic for subclasses of task into a factory class

Abstracted logic (to instantiate a TODO,DEADLINE,EVENT) into a separate factory class.

Made Task an abstract class to better adhere to OOP principles

lywich committed:
Style fixes

Style fixes to better adhere to style conventions.

lywich committed:
Added exception handling for delete

Current situation: User was able to input a value larger than the number of tasks to delete, causing an index out of bounds error.

Added an if statement to check if the value added is invalid.

Updated automatic UI testing to test fro this case.

lywich committed:
Renaming class files to better adhere to style convention

* TODO has been renamed to Todo.
* Response has been renamed to Responses.

lywich committed:
Added support to save/read from a .txt file

Current situation: code only reads input and has no way to store info.

Added the ability to read from /data/Connor.txt. This meant that data is shared among sessions so it functions like a bot.

Changes:
* overloaded instructors for tasks (and subclasses) to make reading from memory cleaner.
* Bug fixes to delete command to now handle invalid deletes.
* Added a delete all function to assist automated testing better.

Let's fix Connor.txt from being created twice separately (due to batch script).

lywich committed:
Update .gitignore

Updated .gitignore to ignore data of the bot as it is not meant to be stored on the repo.

lywich committed:
Added support for Date and time

Current situation: Date and time are strings and have no interactions

Added support to make date/year/time interactable, for future use to do sorting and other functions.

Input is assumed to be valid (2020-05-05 1800).

Let's:
* handle the cases like "2020-5-5"
* update automated test.

lywich committed:
Merge pull request #1 from lywich/branch-Level-7

Added support to save/read from a .txt file

lywich committed:
Merge branch 'master' into branch-Level-8

lywich committed:
Merge pull request #2 from lywich/branch-Level-8

Branch level 8

lywich committed:
Fixing merge error

Current situation: merged master into branch causing the code to break.

Fixed merge error, code should now function as intended.

lywich committed:
Merge pull request #3 from lywich/branch-Level-8

Fixing merge error

lywich committed:
Added Enum to ensure only valid commands are used

Added Enums with valid commands to ensure only valid commands get parsed.

Let's
* refactor the exceptions; InvalidCommandException may no longer be needed.

lywich committed:
Changed the indentation style of switch statement to fit convention

Updated style convention of switch statement to better adhere to style guide.

lywich committed:
Removed unused InvalidCommandException

InvalidCommandException is no longer caught as IllegalArgumentException fufills this purpose.

lywich committed:
Added more abstraction to better adhere to goals

Name changes:
* Responses -> Ui
* TaskFactory -> Parser

Abstracted store/retrieve function to storage from TaskList.

Noticed some redundancy with Todo/Event/Deadline constructors.

Constructors no longer take a LocalDateTime and dataformat;
Instead the raw input String is taken and parsed.

Let's handle corrupted data and invalid date/time.
Current code adds a task before validating a valid DateTime, causing errors with corrupt data or reading invalid data.

lywich committed:
Reorganized files into packages

Reorganized files into packages to make it easier to track different files.

connor.task contains:
* Task, Todo, Deadline, Event
* Invalid Task Exception
* TaskList.

connor.parser contains:
* Parser.

connor.storage contains:
* CorruptedDataException
* Storage.

connor.ui contains:
* Ui.

connor contains:
* Connor.

lywich committed:
Abstracted print messages from TaskList to Ui

Current situation: TaskList contains many print statements that can be better abstracted to Ui.

This is to better adhere to OOP principles as Ui is meant to be the interface for the user to see messages.

lywich committed:
Update .gitignore

lywich committed:
Merge branch 'master' into add-gradle-support

lywich committed:
Merge pull request #4 from lywich/add-gradle-support

Add gradle support

lywich committed:
Update build.gradle

lywich committed:
Refactoring of code to make it more OOP

Current situation: instances of Storage and Tasklist had Ui in its constructor.

TaskList had both a Storage and Ui, and Sotrage had Ui again.

This was not ideal so this was removed as it did not adhere to OOP principle.

Code was refactored and methods utilising Ui was changed to continue to function as per normal.

lywich committed:
Some Junit testcases for TaskList

Added a few Junit testcases for TaskList.

Current problem is that Junit fails when testing for inputs due to line separators.

lywich committed:
Junit testcase for Task

One testcase completed for TaskTest, more testcases will be committed.

lywich committed:
Added testcases for Task and TaskList

Added JUnit testcases to test for all public methods in Task and Tasklist

lywich committed:
Update EXPECTED.TXT

Updated expected.txt to reflect changes to .toString method in TaskList.

lywich committed:
Added find functionality

Current situation: there is no way to find a keyword in the list.

Find functionality is a very helpful function to use for this bot.

Added find functionality in TaskList to print every single tasks that has the keyword in their taskname.

This meant that finding "work" will find "homework", this is an intended effect.

lywich committed:
Revert "Added find functionality"

This reverts commit 4ff6f120111c3909ca8df441e5a5cdd96ce335d9.

lywich committed:
Added find functionality

Current situation: there is no way to find a keyword in the list.

Find functionality is a very helpful function to use for this bot.

Added find functionality in TaskList to print every single tasks that has the keyword in their taskname.

This meant that finding "work" will find "homework", this is an intended effect.

lywich committed:
Added JavaDoc

JavaDoc added for Parser, Task and some methods of Storage.

lywich committed:
Added JavaDoc for all non trivial methods in src.main

lywich committed:
Coding standard update to better adhere to coding convention

Removed an unnecessary new line to adhere to coding convention and improve readability.

lywich committed:
Merge pull request #5 from lywich/branch-A-CodingStandard

Coding standard update to better adhere to coding convention

lywich committed:
Merge branch 'master' into branch-Level-9

lywich committed:
Merge pull request #6 from lywich/branch-Level-9

Added find functionality

lywich committed:
Merge pull request #7 from lywich/branch-A-JavaDoc

Branch a java doc

lywich committed:
Resolved merge conflicts and touch ups

Removed duplicate find in TaskList due to merge error.

Removed redundant imports and class variables in TaskListTest.

lywich committed:
Added JUnit test for find in TaskList

lywich committed:
Added javaDoc for parameters in Task

lywich committed:
Added Checkstyle to Gradle

Added Checkstyle capabilities to Gradle.

Fixed errors that violated Checkstyle at the same time.

lywich committed:
Checkstyle violation fix

Removed trailing whitespace.

lywich committed:
Added skeleton GUI

GUI tutorial first 3 steps.

Will be looking towards personalizing the GUI

bofeng1999 commented on src/main/java/connor/ui/Ui.java:
```diff
@@ -0,0 +1,58 @@
+    public static final String LINE = "        ________________________________________________________\n";
+
+    public static void printMessage(String response) {
+        System.out.println(LINE + "        " + response);
```
I feel that the empty spaces can be extracted as a constant variable and appended to the output strings, instead of adding the spaces in front like this. May have inconsistent number of spaces too!

bofeng1999 commented on src/main/java/connor/ui/Ui.java:
```diff
@@ -0,0 +1,58 @@
+
+    public void addTaskMessage(Task task, int size) {
+        String message = "I have added " + task.getTaskName() + " to my memory\n";
+        message = message + "          " + task.toString() + "\n";
```
Same as above, such spaces can be a variable instead

bofeng1999 commented on src/main/java/connor/task/Task.java:
```diff
@@ -0,0 +1,132 @@
+     * Returns String representing if a task is marked.
+     *
+     * @return String [X] if marked.
+     */
```
Comments for getters/setters may not be necessary

bofeng1999 commented on src/main/java/connor/storage/Storage.java:
```diff
@@ -0,0 +1,98 @@
+     */
+    public Task interpretLine(String str) throws CorruptedDataException {
+        String[] directives = str.split("\\|");
+        if (directives[0].equals("T")) {
```
Maybe a switch statement can be used instead of if-else if

bofeng1999 reviewed (commented):
LGTM! Just some minor comments for possible improvements :)

lywich committed:
Added GUI functionality

Program now has a basic GUI functionality that interprets user commands.

Abstracted the enum class Command to a separate class in the parser package to improve readability.

Let's
* abstract GUI methods to a separate GUI class.
* clean up obsolete methods/functions that came with textUIs.

lywich committed:
Merge pull request #8 from lywich/branch-Level-10

Branch level 10

lywich committed:
Added Assertions to Task and TaskList.

The current Task and TaskList rely on some assumptions:
* The name is valid
* The task to be marked/unmarked is valid.

This dangerous as this will throw an error/unintended behaviour.

Added assertions to ensure:
* Taskname will never be empty.
* Users can only mark valid tasks.

Although these are already caught somewhere, this is to make one last safety net.

Let's,
* identify more assumptions made.

lywich committed:
Removed non-essential methods

Cleaned up mistakes that came from past refactoring.

lywich committed:
Abstracted the logic for GUI to a separate class.

Currently, Connor.java contains both the GUI and Connor methods.

Not only is this messy, both serve different functions.

Abstracted javaFX methods and methods to separate class in gui package.
Merged the functions of Launcher into Connor.

Launcher no longer serves a purpose can be merged with Connor.
Connor now "launches" the gui.

Let's abstract away details from the main method in Gui.java.

lywich committed:
Abstracted logic from main method.

Separated logic from main method to improve readability.

lywich committed:
Merge pull request #9 from lywich/branch-A-Assertions

Added Assertions to Task and TaskList.

lywich committed:
Merge pull request #10 from lywich/branch-A-CodeQuality

Branch a code quality

lywich committed:
Added Find functionality

Connor bot has no find functionality.

Added find functionality.
Refactored the code to accomodate this new function.

Let's refactor the code,
* Added compareTo to each subclasses to enable sorting
* Added a new sort command, which is called in TaskList.
* Sorting utilizes the LocalDateTime compareTo, required refactoring.
* LocalDateTime is now passed into constructors to do so.

Let's implement sorting via this way:
* Deadline is sorted with the earliest timing first.
* Event is sorted by earliest start timing, then earliest end timing.
* If any of above no longer apply, sort by taskName alphabetically.
* Lastly, tasks are sorted by todo first, deadline, then Event.

lywich committed:
Merge pull request #11 from lywich/branch-BCD-Extension

Added Find functionality

lywich committed:
Updated user image to Hank

lywich committed:
Create gradle.yml

lywich committed:
Create suppressions.xml

lywich committed:
Revert "Create suppressions.xml"

This reverts commit d5e92ec9dc00c17fe6ad7193f320e9fb24d12707.

lywich committed:
Fixed broken checkstyle

Checkstyle was broken; Throws "unable to create root module" error.
This was found out when gradle.yml was added.

Changes:
* Added Suppressions.xml to fix the problem.
* Fixed checkstyle errors found in the project.

lywich committed:
Update gradle.yml

Removed IO redirection test for the project as feature has been depreciated.

lywich committed:
Create suppressions.xml

lywich committed:
Improved GUI

GUI lacks polish.

Added background color to both Hank and Connor.

lywich committed:
Added User Guide

Updated readme to reflect on the features of the program.

Added product screenshot.

Added the function to exit program after the command "bye".

lywich committed:
Update TaskListTest.java

Updated "find" testcase to reflect change to method.

lywich committed:
Create MANIFEST.MF

lywich committed:
Final clean up for release

Fixed grammar mistakes in README.
Added javadoc for newly added methods.
Added some readability improvements.

lywich committed:
Fixed readme

Added more acknowledgement for pictures used.
Changed the picture format of Ui.png.

lywich committed:
Updated readme and fixed minor bug

Updated readme to be easier to understand.

Caught the exception whereby an invalid date is parsed.

lywich committed:
Updated readme formatting

Updated readme and removed unused import in Parser.

lywich committed:
Revert "Updated readme formatting"

This reverts commit c398f128c103c9a6573284e4ead8b8e725884be6.

lywich committed:
Update readme formatting

Update readme formatting.
Remove unused import in Parser.

lywich committed:
Updated readme formatting

Updated readme formatting.
Removed unused import in Parser.

lywich committed:
Merge branch 'master' of https://github.com/lywich/ip

lywich committed:
Fix style issues and remove unneeded import

Code calls unneeded Collections.sort in TaskList.java.

This is unneeded as:
* It requires an import
* Can use this.tasks as it is already a list.

Sort utilizes the List.sort rather than Collections.sort.

Fixed style issues:
* message printing to a single line for readability
* added override to compareTo in Task.

lywich committed:
Update README.md

lywich committed:
Update README.md

lywich committed:
Update README.md

lywich committed:
Update README.md

lywich committed:
Update README.md

lywich committed:
Update README.md

lywich committed:
Update README.md

lywich committed:
Update README.md
